### PR TITLE
Detect client disconnection while running query and interrupt execution of query

### PR DESCRIPTION
**Pull request by maksm90**
### Patch description

This PR is in fact the backport of [similar commit](https://git.postgresql.org/gitweb/?p=postgresql.git;a=commit;h=c30f54ad732ca5c8762bb68bbe0f51de9137dd72) from postgres core with some additions and [some depending change](https://git.postgresql.org/gitweb/?p=postgresql.git;a=commit;h=22f6f2c1ccb56e0d6a159d4562418587e4b10e01) in timeout module. It provides a new GUC `check_client_connection_interval` that can be used to periodically check via `CLIENT_CHECK_CONNECTION_TIMEOUT` interrupts whether the client connection has gone away, while running very long queries. Now, it only works for client-coordinator interaction and client-segment in utility mode. The patch performs non-locking check of socket state using non-standard Linux extension (also adopted by at least one other OS) - POLLRDHUP option that is not defined by POSIX. For Darwin, it relies on regular POSIX-compliant POLLHUP option.

Greenplum already has the [similar feature](https://github.com/greenplum-db/gpdb/blob/7.0.0-alpha.0/src/backend/cdb/motion/ic_udpifc.c#L5783-L5801) implemented inside the loop of sending/receiving chunks of rows via UDP interconnect. But this implementation cannot be considered satisfactory: 

1. It relies on [recv(..., ..., MSG_PEEK | MSG_DONTWAIT)](https://github.com/greenplum-db/gpdb/blob/7.0.0-alpha.0/src/backend/cdb/motion/ic_udpifc.c#L5759-L5778) syscall for Unix-like systems that doesn't account pipelined kind of connection termination (when `X` message about connection termination arrives to incoming socket buffer while the postgres process is not able to handle it and `recv(MSG_PEEK | MSG_DONTWAIT)` call doesn't recognize further connection termination).
2. The logic might be unreachable within the satisfactory time interval when the node performs some heavy processing (e.g., long `pg_sleep()` call).
3. Connection checking is activated when the node interacts with others via UDP interconnect, thereby, we exclude the cases when query is executed on single node and interconnect has other than UDP types.

Current implementation hasn't these drawbacks. Old implementation for client-dispatcher interaction was removed.

### Test scenarios

Current patch was fully tested on three test cases on Linux and MacOS platforms:

1. Killing of client process.
    In this case client's OS sends FIN message and server have to handle changed socket state on the next iteration of `CLIENT_CONNECTION_CHECK_TIMEOUT` interrupt triggering.
2. Emulation cable breakdown.
    This case requires to set settings related with keepalive behavior (`tcp_keepalives_idle`, `tcp_keepalives_interval` and `tcp_keepalives_count`). And server terminates "zombie" backend after connection reset ensured by keepalive procedure.
 3. Graceful closing connection by client asynchronously executing long query.
    After asynchronous sending long query client have to close connection by sending 'X' message (calling `PQfinish` libpq function). Hereon, connection is gracefully closed by client and backend process have to cancel query execution on the next iteration of `CLIENT_CONNECTION_CHECK_TIMEOUT` interrupt.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/backend/tcop/postgres.c [hunk 1]
```diff
@@ -3328,6 +3331,16 @@ start_xact_command(void)
...
+
+	/* Start timeout for checking if the client has gone away if necessary. */
+	if (client_connection_check_interval > 0 &&
+		/* doesn't operate on segments in case of distributed queries */
+		Gp_role != GP_ROLE_EXECUTE &&
+		IsUnderPostmaster &&
+		MyProcPort &&
+		!get_timeout_active(CLIENT_CONNECTION_CHECK_TIMEOUT))
+		enable_timeout_after(CLIENT_CONNECTION_CHECK_TIMEOUT,
+							 client_connection_check_interval);
```
Comment (interma) [inline src/backend/tcop/postgres.c:3343]: many conditions here, just a reminder: maybe need to check whether missing others.

**Comment by interma** [comment]
I reviewed this PR roughly, looks fine: most of the code is the same as postgres:
https://git.postgresql.org/gitweb/?p=postgresql.git;a=commit;h=c30f54ad732ca5c8762bb68bbe0f51de9137dd72
https://git.postgresql.org/gitweb/?p=postgresql.git;a=commit;h=22f6f2c1ccb56e0d6a159d4562418587e4b10e01

I will setup this PR in my test env and verify it later.

File: src/backend/cdb/motion/ic_udpifc.c [hunk 2]
```diff
@@ -5802,16 +5802,12 @@ dispatcherAYT(void)
 static void
 checkQDConnectionAlive(void)
 {
-	if (!dispatcherAYT())
+	if (Gp_role == GP_ROLE_EXECUTE)
```
Comment (interma) [inline src/backend/cdb/motion/ic_udpifc.c:5805]: looks no need to check at QD side.
But give a warning is better, just in case.

**Comment by interma** [comment]
@maksm90 one question about your comment:
> Old implementation for client-dispatcher interaction was removed

Do you mean it's the modification of `checkQDConnectionAlive()`? But it's a static func and only used in udfipc (between QD and QE). So I think you don't touch any previous "implementation for client-dispatcher interaction" logic, right?

File: src/backend/tcop/postgres.c (hunk 1 above)
Comment (maksm90) [reply src/backend/tcop/postgres.c:3343]: I'm afraid I'll have to leave all conjuncts trying to get the logical complement of this expression. Furthermore, this condition was not so changed compared with [original patch](https://github.com/postgres/postgres/commit/c30f54ad732ca5c8762bb68bbe0f51de9137dd72#diff-0e44d172203e2969b2440b26d5a272c39e3dcef3013351e3357c32a97a7b7944R2678-R2682) to vanilla postgres. I've just added conjunct verifying the role of current node.

File: src/backend/cdb/motion/ic_udpifc.c (hunk 2 above)
Comment (maksm90) [reply src/backend/cdb/motion/ic_udpifc.c:5805]: Maybe the name `checkQDConnectionAlive` is not so appropriate here but the logic above is executed on any side of interconnect communication (QD and QEs) under receiving packets (`receiveChunksUDPIFC()` call) and sending their (`SendChunkUDPIFC()` | `SendEosUDPIFC()` -> `checkExceptions()`).
By now, I have removed the legacy check connection logic where we have a straight interaction with client side, i.e. on QD side or when query is executed in utility mode.
Later I'll plan to expand current patch to cover verification of connection liveness between QEs and QEs with QD and then will remove all this block.

**Comment by maksm90** [comment]
Hi @interma ! Thanks for your attention to this work.

Now in Arenadata we have figured out that current implementation of this patch is not safe to use - we have observed unpleasant error emitting when handling of `CLIENT_CONNECTION_CHECK_TIMEOUT` event occurs inside `CHECK_FOR_INTERRUPTS()` call. Perhaps, it's worth to move connection check logic directly into `ClientCheckTimeoutHandler()`. Details of this will be presented later in a separate new issue.

> Do you mean it's the modification of checkQDConnectionAlive()? 

Yes.

> But it's a static func and only used in udfipc (between QD and QE). So I think you don't touch any previous "implementation for client-dispatcher interaction" logic, right?

Excluding `GP_ROLE_DISPATCH` and `GP_ROLE_UTILITY` roles from body of `checkQDConnectionAlive()` makes legacy connection checking routine available just to QEs, i.e. using `checkQDConnectionAlive()` we check connection only between QEs and QD, not QD and client.

**Comment by interma** [comment]
> Now in Arenadata we have figured out that current implementation of this patch is not safe to use - we have observed unpleasant error emitting when handling of `CLIENT_CONNECTION_CHECK_TIMEOUT` event occurs inside `CHECK_FOR_INTERRUPTS()` call. Perhaps, it's worth to move connection check logic directly into `ClientCheckTimeoutHandler()`. Details of this will be presented later in a separate new issue.

Thanks @maksm90  for your infromnotification and explanation!
So, do you want to close this PR or convert it to "draft" status (pending now and do more commits in future) ?

**Comment by maksm90** [comment]
Hello @interma ! I have posted the issue https://github.com/greenplum-db/gpdb/issues/13048 with which current task is related.

>  So, do you want to close this PR or convert it to "draft" status (pending now and do more commits in future) ?

I select the second option. The straightforward solution is to move client check connection logic into `ClientCheckTimeoutHandler()` function but I would like to fix issue with `InterruptPending` flag as proposed in https://github.com/greenplum-db/gpdb/issues/13048 to leave things as they are. I suppose it's more correctly in general.

**Comment by z-wenlin** [comment]
Any updates for this PR? Thanks.

**Comment by maksm90** [comment]
> Any updates for this PR? Thanks.

I'll remove draft status after merging https://github.com/greenplum-db/gpdb/pull/13148 . Any changes in current patch is not required.

**Comment by interma** [comment]
@maksm90 https://github.com/greenplum-db/gpdb/pull/13148 has been merged, so I am preparing to merge this one.

One more question: does this PR need to backport to 6x?

**Comment by ivannovick** [comment]
thx folks, great to get this issue resolved.

**Comment by z-wenlin** [comment]
There are some conflicts that need to be resolved before merged. Besides, does the PR for 6X has been raised? 
Thanks.

**Comment by wenwangBJ** [comment]
Hi, @maksm90 It will be very useful for GPDB 6 if we can have the PR for 6X.

**Comment by maksm90** [comment]
> There are some conflicts that need to be resolved before merged.

Resolved, thx.

> Hi, @maksm90 It will be very useful for GPDB 6 if we can have the PR for 6X.

[Done](https://github.com/greenplum-db/gpdb/pull/13285)

**Comment by maksm90** [comment]
It should be noted that currently on master branch of PostgreSQL Tomas Munro [rewrote](https://github.com/postgres/postgres/commit/cba5b994c990bba8df9b8bb75f25ca40aef6b68b) the internal implementation of connection status checking to expand portability of this feature. I considered an option to port this change to current PR but eventually refused this because it requires to backport much dependent changes [[1](https://github.com/postgres/postgres/commit/50e570a59e7f86bb41f029a66b781fc79b8d50f1), [2](https://github.com/postgres/postgres/commit/815c2f0972c8386aba7c606f1ee6690d13b04db2)]. All of these eventually will come to project as long as gpdb will merge with newest versions of PostgreSQL.

**Comment by ashwinstar** [comment]
> It should be noted that currently on master branch of PostgreSQL Tomas Munro [rewrote](https://github.com/postgres/postgres/commit/cba5b994c990bba8df9b8bb75f25ca40aef6b68b) the internal implementation of connection status checking to expand portability of this feature. I considered an option to port this change to current PR but eventually refused this because it requires to backport much dependent changes [[1](https://github.com/postgres/postgres/commit/50e570a59e7f86bb41f029a66b781fc79b8d50f1), [2](https://github.com/postgres/postgres/commit/815c2f0972c8386aba7c606f1ee6690d13b04db2)]. All of these eventually will come to project as long as gpdb will merge with newest versions of PostgreSQL.

Thank you for notifying on this. Given its mostly only for portability aspect and GPDB's major deployment is on linux OS only seems not worth necessary pain to backport this and can wait till we catch-up to latest.

**Comment by interma** [comment]
Thanks @maksm90.
Merge this pr now and I will review the 6x pr later.

**Comment by kainwen** [comment]
@maksm90  @interma 

I see the code:

```c
	 */
	enable_statement_timeout();

	/* Start timeout for checking if the client has gone away if necessary. */
	if (client_connection_check_interval > 0 &&
		/* doesn't operate on segments in case of distributed queries */
		Gp_role != GP_ROLE_EXECUTE &&
		IsUnderPostmaster &&
		MyProcPort &&
		!get_timeout_active(CLIENT_CONNECTION_CHECK_TIMEOUT))
		enable_timeout_after(CLIENT_CONNECTION_CHECK_TIMEOUT,
							 client_connection_check_interval);
}
```

I have a question: why do we exclude QEs here? 

Thanks a lot!

**Comment by interma** [comment]
> I have a question: why do we exclude QEs here?

My thoughts: 
* the scenario of this PR just focuses on detecting the quit event in the "client side" (e.g. psql, jdbc connection).
* most of scenarios which QD quit will send cancel message to QE's: force them to quit.
* specific scenarios, e.g. `kill -9` QD, **do client_connection_check() on QE may help**, we can do an investigation later.

Looks the key point is: "client side" is out of GP, we cannot control them, so need to detect. But QD is a GP component, we assume it doesn't do an abnormal exit often.

**Comment by kainwen** [comment]
> > I have a question: why do we exclude QEs here?
> 
> My thoughts:
> 
> * the scenario of this PR just focuses on detecting the quit event in the "client side" (e.g. psql, jdbc connection).
> * most of scenarios which QD quit will send cancel message to QE's: force them to quit.
> * specific scenarios, e.g. `kill -9` QD, **do client_connection_check() on QE may help**, we can do an investigation later.
> 
> Looks the key point is: "client side" is out of GP, we cannot control them, so need to detect. But QD is a GP component, we assume it doesn't do an abnormal exit often.

Thanks for the reply.

Normally, even for `pg_terminate_backend` the QD must exit finally, it will destroy dispatch state, and recycle each gang. If the Gang is not idle, currently GPDB will signal each QE to force cancel, this progress is in serial. And cancel one QE involve: (1) connect to the postmaster (2) send a packet (3) postmaster fork a worker to receive the package and send signal to QE (4) reply to QD.

In high workload and huge cluster, the above serial progress may take a lot of time to cancel QEs.

Since `pg_terminate_backend`, the QD finally must exit, I think we might not want to spend so much time to quit each QE using the network ping-pong.

**Comment by maksm90** [comment]
> why do we exclude QEs here?

Hi @kainwen!
Initially I treated to stay the logic of terminating backend on QE when QD lost its libpq connection to it. But this functionality requires some additional reworks. For example, we might end up to non-terminated state of backend process in the following scenario:
1. Initialize database:
```bash
gpconfig -c client_connection_check_interval -v '1s'
gpconfig -c tcp_keepalives_interval -v '5s'
gpconfig -c tcp_keepalives_idle -v '1s'
gpconfig -c tcp_keepalives_count -v 2
gpstop -u
```
2. Run simple distributed long running query and isolate some segment:
```sql
create table test as select i from generate_series(1, 1);

# create simple identity function with huge delay for attribute datatype
CREATE OR REPLACE FUNCTION public.same_with_delay(i integer)
 RETURNS integer
 LANGUAGE plpgsql
AS $function$
begin
    perform pg_sleep(1000);
    return i;
end;
$function$;

select count(1) from test where same_with_delay(i) = 0;
```
```shell
iptables -A INPUT -i lo -p tcp --destination-port SEGMENT_PORT -j DROP; \
iptables -A OUTPUT -o lo -p tcp --source-port SEGMENT_PORT -j DROP
```
3. After around 7-10 seconds backend on segment terminates with the following error: `connection to client lost`. And backend on master hangs reading row from QE. Backtrace of master backend process is:
```C
#0  epoll_wait (epfd=28, events=0x55a9ec2a4f18, maxevents=3, timeout=250) at ../sysdeps/unix/sysv/linux/epoll_wait.c:30
#1  WaitEventSetWaitBlock (set=0x55a9ec2a4ea0, cur_timeout=250, occurred_events=0x55a9ec151bd0, nevents=3) at latch.c:1081
#2  WaitEventSetWait (set=0x55a9ec2a4ea0, timeout=250, occurred_events=0x55a9ec151bd0, nevents=3, wait_event_info=134217770) at latch.c:1033
#3  receiveChunksUDPIFCLoop (pTransportStates=0x55a9ec3795a8, pEntry=0x55a9ec443960, srcRoute=0x7ffc30bd5590, conn=0x0, waitset=0x55a9ec2a4ea0, nevent=3) at ic_udpifc.c:3812
#4  receiveChunksUDPIFC (pTransportStates=0x55a9ec3795a8, pEntry=0x55a9ec443960, motNodeID=1, srcRoute=0x7ffc30bd5590, conn=0x0) at ic_udpifc.c:3724
#5  RecvTupleChunkFromAnyUDPIFC_Internal (transportStates=0x55a9ec3795a8, motNodeID=1, srcRoute=0x7ffc30bd5590) at ic_udpifc.c:3933
#6  RecvTupleChunkFromAnyUDPIFC (transportStates=0x55a9ec3795a8, motNodeID=1, srcRoute=0x7ffc30bd5590) at ic_udpifc.c:3953
#7  processIncomingChunks (mlStates=0x55a9ec1519d0, transportStates=0x55a9ec3795a8, pMNEntry=0x55a9ec151a28, motNodeID=1, srcRoute=-100) at cdbmotion.c:671
#8  RecvTupleFrom (mlStates=0x55a9ec1519d0, transportStates=0x55a9ec3795a8, motNodeID=1, srcRoute=-100) at cdbmotion.c:624
#9  execMotionUnsortedReceiver (node=0x55a9ec3740c0) at nodeMotion.c:351
#10 ExecMotion (pstate=0x55a9ec3740c0) at nodeMotion.c:152
...
```
4. Explicit termination of master backend imposes permanent hanging on cancellation of slice process on isolated segment.
```shell
kill -TERM MASTER_BACKEND_PID
```
```C
#0  __libc_connect (fd=29, addr=..., len=16) at ../sysdeps/unix/sysv/linux/connect.c:26
#1  internal_cancel (raddr=0x55a9ec29e1a0, be_pid=683747, be_key=-973239885, errbuf=0x7ffc30bd4dc0 "", errbufsize=256, requestFinish=false) at fe-connect.c:4392
#2  PQcancel (cancel=0x55a9ec29e1a0, errbuf=0x7ffc30bd4dc0 "", errbufsize=256) at fe-connect.c:4516
#3  cdbconn_signalQE (segdbDesc=0x55a9ec2b1618, errbuf=0x7ffc30bd4dc0 "", isCancel=true) at cdbconn.c:472
#4  cdbconn_disconnect (segdbDesc=0x55a9ec2b1618) at cdbconn.c:328
#5  cdbconn_termSegmentDescriptor (segdbDesc=0x55a9ec2b1618) at cdbconn.c:110
#6  cdbcomponent_recycleIdleQE (segdbDesc=0x55a9ec2b1618, forceDestroy=false) at cdbutil.c:951
#7  RecycleGang (gp=0x55a9ec3b9880, forceDestroy=false) at cdbgang.c:944
#8  cdbdisp_destroyDispatcherState (ds=0x55a9ec3b9828) at cdbdisp.c:407
#9  cleanup_dispatcher_handle (h=0x55a9ec3b97f0) at cdbdisp.c:531
#10 cdbdisp_cleanupDispatcherHandle (owner=0x55a9ec12c3b0) at cdbdisp.c:595
#11 CdbResourceOwnerWalker (owner=0x55a9ec12c3b0, callback=0x55a9e3b98cb5 <cdbdisp_cleanupDispatcherHandle>) at resowner.c:1380
#12 CdbResourceOwnerWalker (owner=0x55a9ec172b30, callback=0x55a9e3b98cb5 <cdbdisp_cleanupDispatcherHandle>) at resowner.c:1384
#13 AtAbort_DispatcherState () at cdbdisp.c:554
#14 AbortTransaction () at xact.c:3485
...
```

IMO by now the state of gpdb is not ready to activate functionality of current PR for QEs.

**Comment by ashwinstar** [comment]
@maksm90 great thanks for the scenario where its challenging to adapt this mechanism for QEs.

For point 3: would be helpful to understand how are these things handled for regular clients (or libpq) in case of connection closed by server due to isolation.

For point 4: What is the behavior without `client_connection_check_interval` on QEs in this isolated segment case? Does the cancellation from master still hang?

**Comment by maksm90** [comment]
> For point 3: would be helpful to understand how are these things handled for regular clients (or libpq) in case of connection closed by server due to isolation.

There's required to setup TCP keepalives on client side to detect connection breakdown when process waits on client socket.

> For point 4: What is the behavior without client_connection_check_interval on QEs in this isolated segment case? Does the cancellation from master still hang?

The same, except that segment backend process doesn't terminate immediately but after the communication stage via libpq socket.

**Comment by maksm90** [comment]
> The same, except that segment backend process doesn't terminate immediately but after the communication stage via libpq socket.

Although, after explicit termination the master backend process eventually completes its work on timeout of connection to segment under `PQcancel()` call. From master log:
```
Unable to cancel: PQcancel() -- connect() failed: Connection timed out,"cdbconn.c",330,
The previous session was reset because its gang was disconnected (session id = 62). The new session id = 68,"cdbgang.c",787,

Anyway, IMO master backend have to detect itself the disconnection to segments and fluently terminates gang processes.
```

**Comment by interma** [comment]
> Anyway, IMO master backend have to detect itself the disconnection to segments and fluently terminates gang processes.

`PQcancel()->internal_cancel()` hard code a 660s poll timeout:
https://github.com/greenplum-db/gpdb/blob/63d53c3c0c013b1d7e2573227d40ca2f8fe8c033/src/interfaces/libpq/fe-connect.c#L4444
So QD hangs long time.

I think a proper way is: set a smaller timeout value when enable `client_connection_check_interval` on QEs, then QD won't wait too long time and let the hung QE quit itself later.

**Comment by maksm90** [comment]
> `PQcancel()->internal_cancel()` hard code a 660s poll timeout:
> 
> https://github.com/greenplum-db/gpdb/blob/63d53c3c0c013b1d7e2573227d40ca2f8fe8c033/src/interfaces/libpq/fe-connect.c#L4444

Thanks for pointing this out.

> I think a proper way is: set a smaller timeout value when enable client_connection_check_interval on QEs, then QD won't wait too long time and let the hung QE quit itself later.

I see the first thing is to allow QD to detect disconnection with segments, e.g., by polling postgres connections to segments alongside with client's one in `pq_check_connection()` function. Then cancel request to disconnected segment might not be required further.

**Pull request merged**